lib/src/solver/result.dart

```diff
@@ -67,27 +93,56 @@ class SolveResult {
             .intersect(sdkConstraints[identifier] ?? VersionConstraint.any);
       });
     }
-
-    return LockFile(packages,
-        sdkConstraints: sdkConstraints,
-        mainDependencies: MapKeySet(_root.dependencies),
-        devDependencies: MapKeySet(_root.devDependencies),
-        overriddenDependencies: MapKeySet(_root.dependencyOverrides));
+    return LockFile(
+      packages.map((id) {
+        var description = id.description;
+        // Use the cached content-hashes after downloading to ensure that
+        // content-hashes from legacy servers gets used.
+        if (description is ResolvedHostedDescription) {
+          Uint8List? cachedHash =
+              description.description.source.sha256FromCache(id, cache);
+          if (cachedHash == null) {
+            // This should not happen.
+            throw StateError(
+                'Archive for ${id.name}-${id.version} has no content hash.');
+          }
+          final originalEntry = _previousLockFile.packages[id.name];
+          if (originalEntry != null && originalEntry.version == id.version) {
+            final originalDescription = originalEntry.description;
+            if (originalDescription is ResolvedHostedDescription) {
+              final Uint8List? originalHash = originalDescription.sha256;
+              if (originalHash != null) {
+                if (!bytesEquals(cachedHash, originalHash)) {
+                  log.warning('''
+Content of ${id.name}-${id.version} has changed compared to what was locked your pubspec.lock.
```

Comment:
```suggestion
Content of ${id.name}-${id.version} has changed compared to what was locked in your pubspec.lock.
```

Reply:
Or perhaps: 

The content of $ {id.name}-${id.version} on the server doesn't match the content of what was locked in your pubspec.lock.

Either way is okay, I think.